frontend/src/app/main/data/workspace/libraries.cljs

```diff
@@ -106,3 +112,477 @@
+                                                     :val nil}
+                                                    {:type :set
+                                                     :attr :shape-ref
+                                                     :val nil}]})
+                                    updated-shapes))]
+
+            (rx/of (dwc/commit-changes rchanges uchanges {:commit-local? true})
+                   (dws/select-shapes (d/ordered-set (:id group))))))))))
+
+(defn- make-component-shape
+  "Clone the shape and all children. Generate new ids and detach
+  from parent and frame. Update the original shapes to have links
+  to the new ones."
+  [shape parent-id objects]
+  (let [xf-new-shape (fn [new-shape original-shape]
+                       (assoc new-shape :frame-id nil))
+
+        xf-original-shape (fn [original-shape new-shape]
+                            (cond-> original-shape
+                              true
+                              (assoc :shape-ref (:id new-shape))
+
+                              (nil? (:parent-id new-shape))
+                              (assoc :component-id (:id new-shape))))]
+
+    (cph/clone-object shape parent-id objects xf-new-shape xf-original-shape)))
+
+(defn delete-component
+  [{:keys [id] :as params}]
+  (us/assert ::us/uuid id)
+  (ptk/reify ::delete-component
+    ptk/WatchEvent
+    (watch [_ state stream]
+      (let [component (get-in state [:workspace-data :components id])
+
+            rchanges [{:type :del-component
+                       :id id}]
+
+            uchanges [{:type :add-component
+                       :id id
+                       :name (:name component)
+                       :new-shapes (:objects component)}]]
+
+        (rx/of (dwc/commit-changes rchanges uchanges {:commit-local? true}))))))
+
+(defn instantiate-component
+  [file-id component-id]
+  (us/assert (s/nilable ::us/uuid) file-id)
+  (us/assert ::us/uuid component-id)
+  (ptk/reify ::instantiate-component
+    ptk/WatchEvent
+    (watch [_ state stream]
+      (let [component (if (nil? file-id)
+                        (get-in state [:workspace-data :components component-id])
+                        (get-in state [:workspace-libraries file-id :data :components component-id]))
+            component-shape (get-in component [:objects (:id component)])
+
+            orig-pos  (gpt/point (:x component-shape) (:y component-shape))
+            mouse-pos @ms/mouse-position
+            delta     (gpt/subtract mouse-pos orig-pos)
+
+            _ (js/console.log "orig-pos" (clj->js orig-pos))
+            _ (js/console.log "mouse-pos" (clj->js mouse-pos))
+            _ (js/console.log "delta" (clj->js delta))
+
+            page-id   (:current-page-id state)
+            objects   (dwc/lookup-page-objects state page-id)
+            unames    (atom (dwc/retrieve-used-names objects))
+
+            all-frames (cph/select-frames objects)
+
+            xf-new-shape
```

Comment:
`xf-` prefix is a convention for transducers, it is confusing when a simple functions uses the same naming